### PR TITLE
Break apart FilterableListControls

### PR DESCRIPTION
**Pull request by anselmbradford**
FilterableListControls contains the controls, notifications, results and pagination. It really should instead have a parent module, FilterableList, that includes those elements. 

## Additions

- Adds FilterableList module to encompass FilterableListControls, Notification, the results (post previews, activity log, or info units), and pagination.

## Removals

- filterable-list-controls.less and moves margin logic to `block` and the JS.
- Removes `filter_type` option (see changes from #4774)

## Changes

- Removes filterable list template `render` macro and calls `value` values directly, like other templates that render directly from the block.
- Update cucumber feature file to remove 0 from dates. 
- Hide result list when there are errors on the client-side.
- Add label to server-side field error message, for consistency with client-side error message (see note section below though for differences that remain).

## Testing

1. Pull branch and run `gulp clean && gulp build`
2. Check the following states.

### No posts state

Example URL: http://localhost:8000/about-us/events/
No content shown.
<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>JS</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>

---

Example URL: http://localhost:8000/about-us/events/
Content for "No posts message" in Wagtail added.
<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td></td>
    <td></td>
    <td></td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>JS</td>
    <td></td>
    <td></td>
    <td></td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>

† Notification shown is in the default gray state.

---

### Page loaded state, no filter interaction.

Example URL: http://localhost:8000/about-us/newsroom/

<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td></td>
    <td></td>
    <td>x</td>
  </tr>
  <tr>
    <td>JS</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td></td>
    <td></td>
    <td>x</td>
  </tr>
</table>
† Is in expanded state.

---

### Page loaded state, single filter applied.

Example URL: http://localhost:8000/about-us/newsroom/?categories=blog

<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td>x</td>
  </tr>
  <tr>
    <td>JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td>x</td>
  </tr>
</table>
† Is in expanded state.

---

### Page loaded state, filters applied, small result set.

Example URL: http://localhost:8000/about-us/newsroom/?categories=blog&topics=financial-education&authors=holly-petraeus

<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>
† Is in expanded state.

---

### Page loaded state, filters applied, no result set.

Example URL: http://localhost:8000/about-us/newsroom/?categories=blog&topics=financial-education&authors=holly-petraeus&from_date=2018

<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>
† Is in expanded state.

---

### Page loaded state, filters applied, erroneous entry (one field).

Example URL: http://localhost:8000/about-us/newsroom/?to_date=asdf

<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td>x†</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>JS</td>
    <td>x†</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>
† Is in expanded state.

---

### Page loaded state, filters applied, erroneous entry (multiple fields).

Example URL: http://localhost:8000/about-us/newsroom/?to_date=asdf&from_date=asdf

<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td>x†</td>
    <td></td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>JS</td>
    <td>x†</td>
    <td></td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>
† Is in expanded state.

---

### Page loaded state, activity-log.

Example URL: http://localhost:8000/activity-log/

<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td></td>
    <td>x</td>
  </tr>
  <tr>
    <td>JS</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td></td>
    <td>x</td>
  </tr>
</table>
† Is in expanded state.

---

### Page loaded state, info-unit.

Example URL: http://localhost:8000/consumer-tools/everyone-has-a-story/

<table>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4">notification</td>
    <td colspan="3">results</td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td>filter control</td>
    <td>single field error</td>
    <td>multi-field error</td>
    <td>⚠</td>
    <td>✓</td>
    <td>post-preview</td>
    <td>activity-log</td>
    <td>info-unit</td>
    <td>pager</td>
  </tr>
  <tr>
    <td>NO-JS</td>
    <td>x†</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td>x</td>
  </tr>
  <tr>
    <td>JS</td>
    <td>x</td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td></td>
    <td>x</td>
    <td>x</td>
  </tr>
</table>
† Is in expanded state.

## Notes

An additional edge-case only-JS behavior is to expand the filter and enter an erroneous value for one of the date filters. It should display an error message and hide the results. It's identical to "Page loaded state, filters applied, erroneous entry (one field)," but performed via JS instead of server-side. However, it differs in the wording of the error message:

Server-side:
screen shot 2019-01-16 at 7 09 52 pm

Client-side:
screen shot 2019-01-16 at 7 11 36 pm

This is due to how the backend and frontend store the label that is added to the error message. This should probably be adjusted to be equal in the future.

### CODE REVIEW

**Comment by schaferjh** [comment]
@anselmbradford :+1: all good.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+   filter_data.page_set: Result set of posts from the filtered content.
+
+   ========================================================================== #}
+
+{% import 'organisms/filterable-list-controls.html' as filterable_list_controls with context %}
+{% import 'molecules/notification.html' as notification with context %}
+
+{# SHOW A NOTIFICATION IF THERE'S NOTHING TO FILTER. #}
+
+{% if value.no_posts_message %}
```
Comment (schbetsy) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: This `if`/`elif` conditional needs to be in the opposite order. Currently, if I fill in the `no_posts_message` field on the "Blog" page, for example (even if there are posts), it acts like there are no posts:

screen shot 2019-01-17 at 10 28 04 am

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            'type': 'default',
+            'is_visible': false,
+            'message': '',
+            'explanation': ''
+        } %}
+
+        {% for field in form %}
+            {% if field.errors %}
+                {% for error in field.errors %}
+                    {% if notification_opts.update({
```
Comment (schbetsy) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Why are these `if` statements empty?

Comment (anselmbradford) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Andy had the same question. Without having them in an if statement the variable was not visible outside of the scope of the for loop. I added the `if` per the suggestion [here](https://stackoverflow.com/questions/9486393/jinja2-change-the-value-of-a-variable-inside-a-loop). Should I add a code comment?

**Comment by chosak** [comment]
> What should be the behavior if a filter has no posts to filter and there has been no message set by content editors?

Should those fields (at least `no_posts_message`, and maybe `no_post_explanation`) be required? I know those were just added in #4758. @schbetsy was it a deliberate design decision to make those optional, or a decision of expediency, as adding them as required would have necessitated adding messages to all existing fields?

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                        block
+                        block__sub
+                        {% if has_active_filters %}
+                        block__flush-top
+                        {% endif %}">
+            {% if value.categories.page_type == 'activity-log' %}
+                {% import 'activity-log/_activity-list.html' as activity_list with context %}
+                {{ activity_list.render(posts) }}
+            {% elif value.output_5050 %}
+                {% from 'molecules/info-unit-2.html' import info_unit with context %}
```
Comment (anselmbradford) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: @Scotchester this should be info-unit-2, right?

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                <div class="o-info-unit-group u-mb30" data-qa-hook="image-text-50-50">
```
Comment (anselmbradford) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: TODO: This whole block should go into its own template, but to manage the complexity of this PR best break things apart here and do a separate PR to separate this part out.

**Comment by schaferjh** [comment]
@anselmbradford would you post a screenshot?

**Comment by schbetsy** [comment]
> @schbetsy was it a deliberate design decision to make those optional

@chosak I didn't see any value in making them required. The only pages where the "No posts message" could reasonably show up are the [events](http://localhost:8000/about-us/events/) and [open notices](http://localhost:8000/policy-compliance/notice-opportunities-comment/open-notices) pages. Making the field required would be pointless for all the other pages, causing confusion and extra work. Having a possible state where nothing shows up doesn't seem like a problem to me.

**Comment by chosak** [comment]
@schbetsy thanks for the explanation, that makes sense.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html (hunk 2 above)
Comment (schbetsy) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: That makes sense 👍. Sure, a comment could be helpful, but isn't 100% necessary.

**Comment by anselmbradford** [comment]
> would you post a screenshot?

@schaferjh 

There are events:
screen shot 2019-01-17 at 11 45 30 am

There are no events and no message setting in Wagtail:
screen shot 2019-01-17 at 11 46 10 am

There are no events and a message is set in Wagtail:
screen shot 2019-01-17 at 11 47 00 am

See Betsy's comment above too as it will reasonably only appear on events and open notices.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html (hunk 1 above)
Comment (anselmbradford) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: blergh, good catch. I pushed up a `has_results` variable and check that.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        } %}
+
+        {% for field in form %}
+            {% if field.errors %}
+                {% for error in field.errors %}
+                    {% if notification_opts.update({
+                        'type': 'error',
+                        'is_visible': true,
+                        'message': notification_opts.message + field.label + ': ' + error + '<br>' | safe
+                    }) %}{%endif%}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]:
```suggestion
                    }) %}{% endif %}
```

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html (hunk 2 above)
Comment (anselmbradford) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Added a comment above each.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+{% elif has_results %}
+
+    <div class="o-filterable-list">
+        {% set form = filter_data.form %}
+        {% set posts = filter_data.page_set %}
+
+        {# ADD THE FILTER CONTROLS. #}
+
+        {% set fragment_id = 'o-filterable-list-controls' %}
+        {{ filterable_list_controls.render(value, form, fragment_id) }}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Why pass in the `fragment_id` instead of hardcoding it in the controls template?

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+            'message': '',
+            'explanation': ''
+        } %}
+
+        {% for field in form %}
+            {% if field.errors %}
+                {% for error in field.errors %}
+                    {# The if statement here allows access to the
+                       notification_opts variable outside of the for loop.
+                    #}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Trivial: I'd either move this close comment tag up to the previous line, or put the opening tag on its own line. (And for other multiline Jinja comments.)

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                    {% if notification_opts.update({
+                        'type': 'error',
+                        'is_visible': true,
+                        'message': notification_opts.message + field.label + ': ' + error + '<br>' | safe
+                    }) %}{%endif%}
+                {% endfor %}
+            {% endif %}
+        {% endfor %}
+        {% if 'filterable-list' in value.form_type and posts is defined %}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Trivial: Consider adding a blank line to visually separate the `endfor` from the `if`.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                {% if count == 0 %}
+                    {# The if statement here allows access to the
+                        notification_opts variable outside of the for loop.
+                    #}
+                    {% if notification_opts.update({
+                        'type': 'warning',
+                        'is_visible': true,
+                        'message': 'Sorry, there were no results based on your filter selections.',
+                        'explanation': 'Please reset the filter or change your selections and try again.'
+                    }) %}{%endif%}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]:
```suggestion
                    }) %}{% endif %}
```

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                {% else %}
+                    {# The if statement here allows access to the
+                        notification_opts variable outside of the for loop.
+                    #}
+                    {% if notification_opts.update({
+                        'type': 'success',
+                        'is_visible': true,
+                        'message': count ~ ' filtered result' ~ count|pluralize('s')
+                    }) %}{%endif%}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]:
```suggestion
                    }) %}{% endif %}
```

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                {% endif %}
+            {% endif %}
+        {% endif %}
+
+        <div class="o-filterable-list_notification
+                    block
+                    block__padded
+                    block__sub
+                    {% if not has_active_filters %}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]:
```suggestion
                    {%- if not has_active_filters %}
```

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+                    u-hidden
+                    {% endif %}">
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]:
```suggestion
                    {%- endif %}">
```

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+                notification_opts.explanation
+            ) }}
+        </div>
+
+        {# ADD THE FILTER RESULTS, IF THERE ARE ANY. #}
+
+        <section class="o-filterable-list_results
+                        block
+                        block__sub
+                        {% if has_active_filters %}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]:
```suggestion
                        {%- if has_active_filters %}
```

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+                        block__flush-top
+                        {% endif %}">
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]:
```suggestion
                        {%- endif %}">
```

**Comment by schbetsy** [comment]
In my opinion, it would make more sense for the filter controls to be open when there is a validation error. The user will have to interact with the expandable to fix the error; why make them press the "+" button to open it?

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+                    {% endif %}">
+            {{ notification.render(
+                notification_opts.type,
+                notification_opts.is_visible,
+                notification_opts.message,
+                notification_opts.explanation
+            ) }}
+        </div>
+
+        {# ADD THE FILTER RESULTS, IF THERE ARE ANY. #}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Can we add an `if` statement here so the empty `section` is not rendered if there are no results?

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+                {% for post in posts %}
+                    {{ post_preview.render(post, value) }}
+                {% endfor %}
+            {% endif %}
+
+            {# DISPLAY THE PAGINATOR, IF THERE ARE RESULTS OVER A COUNT. #}
+
+            {% set total_pages = posts.paginator.num_pages %}
+            {% set current_page = posts.number %}
+            {% if total_pages > 1 and current_page <= total_pages %}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: This condition seems superfluous?

```suggestion
            {% if total_pages > 1 %}
```

**Comment by anselmbradford** [comment]
> In my opinion, it would make more sense for the filter controls to be open when there is a validation error. The user will have to interact with the expandable to fix the error; why make them press the "+" button to open it?

@schbetsy This is fixed in https://github.com/cfpb/cfgov-refresh/pull/4771/commits/f0a18257a15a20ec910aa93f531ad9dfef45b9a6, I thought `has_active_filters` was coming through `value`, but it's on it's own. Should it be global like that or should it be attached to the `value` on the backend?

**Comment by anselmbradford** [comment]
> It's okay for the page to render with the filter components and blankness in this case. It will be a spur to the content editor to remember to set a message in the optional field. Doesn't need to be programmatically decided given it's rarity.

@schaferjh So the other option is to have a default message, but the issue there is what the text would say that would cover the possible cases for the no results.

The problem with completely blank from a frontend perspective is that we can't always assume that a filterable list will occupy space on the page, so have to be aware that there's an edge-case in regard to padding in the layout. Currently with it empty an empty div is added to the page markup that occupies space.

screen shot 2019-01-17 at 2 12 41 pm

File: cfgov/v1/atomic_elements/organisms.py [hunk 17]
```diff
@@ -916,9 +913,10 @@ class FilterControls(BaseExpandable):
     class Meta:
         label = 'Filter Controls'
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply cfgov/v1/atomic_elements/organisms.py]: Updated and didn't need a migration!

File: cfgov/v1/atomic_elements/organisms.py [hunk 18]
```diff
@@ -865,10 +865,7 @@ class Meta:
         classname = 'block__flush-top'
 
 
-# TODO: FilterControls/Filterable List should be updated to use same
-#       atomic name used on the frontend of FilterableListControls,
-#       or vice versa.
-class FilterControls(BaseExpandable):
+class FilterableList(BaseExpandable):
```
[the remark this answers is not in the document]
Comment (anselmbradford) [reply cfgov/v1/atomic_elements/organisms.py]: Filters should be untangled from expandables in multiple ways. I think that's out of scope here, but duly noted!

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html (hunk 1 above)
Comment (anselmbradford) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Updated to `has_unfiltered_results` to better describe it since it had a value when the filters were applied and the warning notification was showing. https://github.com/cfpb/cfgov-refresh/pull/4771/commits/b322bf00c90571c5ca021608bd29adc5308fe4e2

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html (hunk 15 above)
Comment (anselmbradford) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Added in https://github.com/cfpb/cfgov-refresh/pull/4771/commits/b322bf00c90571c5ca021608bd29adc5308fe4e2

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html (hunk 16 above)
Comment (anselmbradford) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Hmmmmm okay I've removed it in https://github.com/cfpb/cfgov-refresh/pull/4771/commits/9086a5fea3d934f9834f91151cadf56f384a4fb0

This is actually a direct duplicate from https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/jinja2/v1/_includes/molecules/pagination.html#L26 that is necessary to hide the pagination container, which is kinda a bummer.

**Comment by Scotchester** [comment]
> This is fixed in [f0a1825](https://github.com/cfpb/cfgov-refresh/commit/f0a18257a15a20ec910aa93f531ad9dfef45b9a6), I thought `has_active_filters` was coming through `value`, but it's on it's own. Should it be global like that or should it be attached to the `value` on the backend?

I like it being separate and leaving `value` reserved for block field values.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html (hunk 16 above)
Comment (Scotchester) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: It seems unnecessary for that to exist there. Seems like it should be up to the thing calling pagination to know if it actually should be calling it or not.

**Comment by schaferjh** [comment]
@anselmbradford hmmm, if that's a problem then @ascott1 might be the better person to weigh in. From a UX perspective, it's such an edge case, I do not see a need to solve for it.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html (hunk 16 above)
Comment (anselmbradford) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Possibly, the method signature looks pretty clear there and it's nice it handles the arguments coming into it correctly.

**Comment by chosak** [comment]
It occurs to me that renaming a block type like this (`FilterControls` to `FilterableList` -- a nice improvement) does require manually running the `block_inventory` Django management command (provided by [wagtail-inventory](https://github.com/cfpb/wagtail-inventory)) to regenerate the page/block list. This powers the "Block Inventory" options in Wagtail admin Settings.

Unfortunately this isn't currently documented in wagtail-inventory, so I've opened https://github.com/cfpb/wagtail-inventory/issues/19 there.

File: cfgov/jinja2/v1/_includes/organisms/filterable-list.html [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+   value: Data object from an instance of the FilterableList StreamField block.
+
+   value.no_posts_message: Message to show in "no posts" state.
+
+   value.no_posts_explanation: More info to show in "no posts" state.
+
+   value.form_type: What type of form this is.
+
+   value.categories.page_type: Page type of the filterable list control.
```
Comment (schbetsy) [inline cfgov/jinja2/v1/_includes/organisms/filterable-list.html]:
```suggestion
   value.categories.page_type: What kind of pages are being filtered, e.g. 'final-rule', 'blog'. 
                               This is used to determine what categories to allow filtering on,
                               as well as how the post preview is displayed.
```

Comment (anselmbradford) [reply cfgov/jinja2/v1/_includes/organisms/filterable-list.html]: Updated in https://github.com/cfpb/cfgov-refresh/pull/4771/commits/dee1b3129ec90e3585884c6f650c6d448aaf72be
I wrapped it a little differently to fit in the 80 character the rest of the comments are in.

**Comment by anselmbradford** [comment]
I defer all JS errors this has introduced to @schbetsy and @Scotchester 😆 Have a good long weekend!
